File: 3b_ADT-example-with-the-Template-Method-Pattern/include/CaffeinatedBeverage.r
#ifndef CAFFEINATED_BEVERAGE_R
#define CAFFEINATED_BEVERAGE_R

#define MAX_CHARS_NAME 20

typedef struct CaffeinatedBeverage_Interface_Struct const * CaffeinatedBeverage_Interface;

typedef struct CaffeinatedBeverage_Data_Struct {
    CaffeinatedBeverage_Interface interface;
    char name[MAX_CHARS_NAME];
} CaffeinatedBeverage_Data_Struct;

typedef struct CaffeinatedBeverage_Interface_Struct {
    void (*brew)(CaffeinatedBeverage_Data_Struct*);
    void (*addCondiments)(CaffeinatedBeverage_Data_Struct*);
    void (*addWhip)(CaffeinatedBeverage_Data_Struct*);
    void (*deinit)(CaffeinatedBeverage_Data_Struct*);
    void (*destroy)(CaffeinatedBeverage_Data_Struct*);
} CaffeinatedBeverage_Interface_Struct;

#endif /*CAFFEINATED_BEVERAGE_R*/

File: 3_Run-time-Polymorphism/e_Design-Pattern_Template-Method/source/CaffeinatedBeverage.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include "assert.h"
#include "CaffeinatedBeverage.h"
#include "CaffeinatedBeverage.r"

void
CaffeinatedBeverage_setName(CaffeinatedBeverage _self, char * name)
{
    ASSERT(_self && name);

    strncpy(_self->name, name, MAX_CHARS_NAME);
}

char *
CaffeinatedBeverage_getName(CaffeinatedBeverage _self)
{
    ASSERT(_self);

    return _self->name;
}

static void CaffeinatedBeverage_boilWater(CaffeinatedBeverage _self __attribute__((unused))) {
    printf("\tBoiling water.\n\r");
}

static void CaffeinatedBeverage_brew(CaffeinatedBeverage _self) {
    ASSERT(_self && _self->interface);

    if( _self->interface->brew ) {
        _self->interface->brew(_self);
    }
}

static void CaffeinatedBeverage_pourInCup(CaffeinatedBeverage _self __attribute__((unused))) {
    printf("\tPouring into cup.\n\r");
}

static void CaffeinatedBeverage_addCondiments(CaffeinatedBeverage _self) {
    ASSERT(_self && _self->interface);

    if( _self->interface->addCondiments ) {
        _self->interface->addCondiments(_self);
    }
}

static void CaffeinatedBeverage_addWhip(CaffeinatedBeverage _self) {
    ASSERT(_self && _self->interface);

    if( _self->interface->addWhip ) {
        _self->interface->addWhip(_self);
    }
}

void CaffeinatedBeverage_prepare(CaffeinatedBeverage _self) {
    ASSERT(_self);

    printf("|__Making beverage with name: %s\n", _self->name);
    CaffeinatedBeverage_boilWater(_self);
    CaffeinatedBeverage_brew(_self);
    CaffeinatedBeverage_pourInCup(_self);
    CaffeinatedBeverage_addCondiments(_self);
    CaffeinatedBeverage_addWhip(_self);
}

void
CaffeinatedBeverage_destroy(CaffeinatedBeverage _self)
{
    ASSERT(_self && _self->interface);

    if( _self->interface->deinit )
    {
        _self->interface->deinit(_self);
    }

    printf("\tDeinitializing caffeinated beverage object with name: %s\n", _self->name);
    memset(_self->name, 0, sizeof(char)*MAX_CHARS_NAME);

    ASSERT( _self->interface->destroy );
    _self->interface->destroy(_self);
}